RFM9x.device.lib.nut

```diff
@@ -0,0 +1,437 @@
+        _writeReg(RFM9X_REG_OP_MODE, mode | (current & 0xf8));
+    }
+        
+    // Clear all interrupt flags
+    function clearInterrupts() {
+        _writeReg(RFM9X_REG_IRQ_FLAGS, 0xff);
+    }
+    
+    // Mask all interrupts. Calling this method will prevent the interrupt
+    // pin from being asserted
+    function maskAllInterrupts() {
+        _writeReg(RFM9X_REG_IRQ_FLAGS_MASK, 0xff);
+    }
+    
+    // Enable a specific interrupt. Avaliable interrupts are in
+    // RFM9X_FLAGS
+    function enableInterrupt(bitnumber) {
+        local currentInterrupts = _readReg(RFM9X_REG_IRQ_FLAGS);
+        // Mask the interrupt by writing high, therefore enable by writing low
+        local newInterrupts = currentInterrupts & (~ (1 << bitnumber));
+        _writeReg(RFM9X_REG_IRQ_FLAGS_MASK, newInterrupts);
+    }
+
+    // Sets the max payload size of the FIFO buffer
+    function setMaxPayload(pl) {
+        pl = pl & 0xff;
+        _writeReg(RFM9X_REG_MAX_PAYLOAD_LENGTH, pl);
+    }
+    
+    // Sets the spreading factor of the radio
+    function setSpreadingFactor(sf) {
+        _writeReg(RFM9X_REG_MODEM_CONFIG2, (sf & 0x0f) << 4);
+    }
+    
+    // Sets the transmission frequency of the radio
+    function setFrequency(freq) {
+        setMode(RFM9X_SLEEP);
+        local reg_freq = (freq/RFM9X_FREQ_STEP).tointeger();
+        local freq_blob = blob(4); // big-endian blob
+        freq_blob[0] = RFM9X_REG_FREQUENCY_HIGH | RFM9X_WRITE_MASK;
+        freq_blob[1] = (reg_freq >> 16) & 0xFF;
+        freq_blob[2] = (reg_freq >> 8) & 0xFF;
+        freq_blob[3] = (reg_freq) & 0xFF;
+        
+        _csLow();
+        _spiModule.writeread(freq_blob);
+        _csHigh();
+    }
+
+    // Sets the receive handler. The receive handler should take two 
+    // parameters: an error message and the received data. The error message
+    // will either be "error" or "valid". If the message is "error", the
+    // received data parameter will be passed as null. 
+    function setReceiveHandler(handler) {
+        _receiveHandler = handler;
+    }
+
+    // Returns whether the last send operation is ongoing
+    function isSending() {
+        return _isSending;
+    }
+    
+    function _setInLoRaMode() {
+       _writeReg(RFM9X_REG_OP_MODE, RFM9X_CONFIG_BYTE);
+    }
+
+    function _configISR() {
+        if (_irqPin) {
+            _irqPin.configure(DIGITAL_IN_PULLDOWN, _interruptServiceRoutine.bindenv(this));
+        } else {
+            throw "interrupt pin passed to constructor is invalid";
+        }
+    }
+
+    function _interruptServiceRoutine() {
+        // Active high
+        if (_irqPin.read()) {
+            local read = _readReg(RFM9X_REG_IRQ_FLAGS);
+            if (read & (1 << RFM9X_FLAGS.PAYLOAD_CRC_ERROR)) {
+                _error = true;
+                _receive("error", null);
+            } else if (read & (1 << RFM9X_FLAGS.TX_DONE)) {
+                _isSending = false;
+                _sendcb("done", null);
```

Comment:
send callback = this is not documented, but it looks like you are using node conventions here, so the first parameter is null if the request was successful otherwise it would be an error message.  This seems like "done" should be passed into the 2nd callback and the first parameter should be null.